lib/screens/v2/candidate_generator/widgets/reconstructed_alert.ex

```diff
@@ -79,22 +79,69 @@ defmodule Screens.V2.CandidateGenerator.Widgets.ReconstructedAlert do
     Enum.member?(@relevant_effects, effect)
   end
 
-  defp relevant_location?(%ReconstructedAlert{alert: alert} = reconstructed_alert) do
+  defp relevant_location?(%ReconstructedAlert{} = reconstructed_alert) do
     case BaseAlert.location(reconstructed_alert) do
       location when location in [:downstream, :upstream] ->
         true
 
       :inside ->
-        alert.effect != :delay or alert.severity > 3
+        relevant_inside_alert?(reconstructed_alert)
 
       location when location in [:boundary_upstream, :boundary_downstream] ->
-        alert.effect != :station_closure and (alert.effect != :delay or alert.severity > 3)
+        relevant_boundary_alert?(reconstructed_alert)
 
       _ ->
         false
     end
   end
 
+  defp relevant_inside_alert?(%ReconstructedAlert{alert: alert} = reconstructed_alert) do
+    alert.effect != :delay or
+      (alert.severity > 3 and relevant_direction?(reconstructed_alert))
+  end
+
+  defp relevant_boundary_alert?(%ReconstructedAlert{alert: alert} = reconstructed_alert) do
+    alert.effect != :station_closure and
+      (alert.effect != :delay or
+         (alert.severity > 3 and relevant_direction?(reconstructed_alert)))
+  end
+
+  # This function assumes that stop_sequences is ordered by direction north/east -> south/west.
+  # If the current station has only one stop sequence and the stop_id is the first or last entry in the stop_sequence,
+  # it is a terminal. Delay alerts heading in the direction of the station are not relevant.
+  defp relevant_direction?(
+         %ReconstructedAlert{
+           screen: %Screen{
+             app_params: %{reconstructed_alert_widget: %CurrentStopId{stop_id: stop_id}}
+           },
+           stop_sequences: [stop_sequence]
+         } = t
+       ) do
+    informed_entities = BaseAlert.informed_entities(t)
+
+    direction_id =
+      informed_entities
+      |> Enum.map(fn %{direction_id: direction_id} -> direction_id end)
+      |> Enum.uniq()
+      |> hd()
```

Comment:
Will an informed entity with a non-nil direction_id always come before any ones with nil direction_id? If not, I think this would be a safer way to always get the first non-nil one, if it exists.
```suggestion
      informed_entities
      |> Enum.find(fn %{direction_id: direction_id} -> not is_nil(direction_id) end)
```
(`Enum.find` returns nil if no matching element is found)

Reply:
I don't _think_ it's possible for `informed_entities` to have nil and non-nil direction ids. So I don't think that should be a problem. I do like leaning on a default, though, so now `hd()` will never cause us issues.